src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentStack.cs

```diff
@@ -624,11 +600,19 @@ private int TryPopCore(int count, out Node poppedHead)
                 }
                 next = head;
                 int nodesCount = 1;
-                for (; nodesCount < count && next._next != null; nodesCount++)
+                for (; nodesCount < count && ReferenceEquals(next, Guard) == false; nodesCount++)
                 {
+                    // fast spin till _next is set
+                    while (next._next == null)
+                    { }
+
                     next = next._next;
                 }
 
+                // fast spin till _next is set
+                while (next._next == null)
```

Comment:
Imagine one-element stack:
`_head: SomeNode`
`SomeNode._next: Guard`

Ask for 2+ elements

Assign `_head` to `next`
`next != Guard`, so enters loop at 603.  Assign `Guard` to `next`, then exits that loop because `next == Guard`.
`next == Guard`, `Guard._next == null`, loop runs forever.

Reply:
Thank you for this case. Will fix it.